Use SmartyStreets geocoding API for addresses

kairstenfay opened the pull request:
EDIT: see latest commit message for updated info.

vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
These changes swapped Google's geocoding API for SmartyStreet's API.
The HTTP GET request requires an authentication token and ID which are
not tracked through git and need to be added in a separate file,
`environment.py`. An example of a valid `environment.py` configuration
is viewable at `example.environment.py`. These authentication strings
are currently imported as global variables in
`address_to_census_tract.py`.

The program currently works by passing SmartyStreets a free text
address via the `street` parameter. There are a couple of assumptions
made about addresses in order for this program to work. Namely,
    1. A street address contains at least one letter, [a-z] (case
    insensitive).
    2. A street address comes before a zip code in an address.

We do not assume that an address is comma-separated.

If an address lookup fails on the first attempt (receives a status code
of 200 but returns no JSON object), then a second attempt is made to
lookup the address. In the second attempt, the program
    1. Splits the address (string) on any commas, converting it into a
    list.
    2. Removes the first index of the list if it contains only numbers.
    For example, in the original address `100, 123 Main St.`, `100`
    would be removed from the list.
    3. Finds and removes a partial string matching the pattern
    `#.*(?\s)` (a pound sign followed by anything with an optional
    space). In the original address `123 Main St. #100`, `#100` would be
    removed.

Removing these unit or suite numbers was essential in getting the
SmartyStreets API to find matches at certain addresses in the
`testset.json` file.

kairstenfay commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Standalone address de-identification program](https://trello.com/c/SQW5WRKs/20-standalone-address-de-identification-program)

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
-from sys import stderr
-
+from sys import stderr, argv
+from environment import AUTH_ID, AUTH_TOKEN  # Global variables
```
Let's make these environment variables instead of Python globals? For example, `os.environ.get("SMARTYSTREETS_AUTH_ID")` instead of importing `AUTH_ID` from an `environment.py`.

kairstenfay replied:
Done and loaded upon conda environment activation, per https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#macos-and-linux

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
+from environment import AUTH_ID, AUTH_TOKEN  # Global variables
+import re
+import requests
+import pandas as pd
```
Do you know about [SmartyStreet's Python SDK](https://smartystreets.com/docs/sdk/python)? We may want to use that instead of rolling our own API client.

kairstenfay replied:
Now using `smartystreets_python_sdk`!

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
-    return geocoder.google(address).latlng
-
+    This is going to prove difficult if addresses are separated by commas.
+    Excel may be the only way to go unless all addresses are wrapped in quotes.
```
The CSV isn't valid if a field value contains unescaped/unquoted commas. I don't think we have to worry about this.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
 
-    with open(geojson_filename) as file:
-        geojson = json.load(file)
+    if 'delivery_line_2' in response.keys():
```
FWIW, the `in` operator works directly against dictionaries, so the `.keys()` is unnecessary.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
+    Find the first polygon in *polygons* (loaded from file) which contains the 
+    *latlng* and return the polygon, else None.
     """
+    polygons = load_geojson("data/geojsons/Washington_2016.geojson")
```
Moving the loading of the geojson into the lat/lng to polygon function means that the geojson is loaded again for every single lookup. That's going to slow things down a bit, so we should lift it up.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
+    true street address.
+
+    The following method should work even if the given address has no commas
+    separating the address components. 
```
I'm not comfortable with the accuracy of trying to parse and reformat addresses ourselves. That's part of reason we're using a third-party API. :-) I wonder why SmartyStreets fails on these? Could it be what we're asking for or the way we're doing a request rather than some fundamental inability? Can we consult with SmartyStreets documentation/support here for a better way to handle the problematic addresses?

What other solutions did you try?

kairstenfay replied:
Your concerns make sense.

Incidentally, switching to the smartstreets-python-sdk library proved more robust for addresses with the Suite or Unit # issue. 

I'm still looking into fixing the lookup for the other problematic address (with unit number out front, plus a comma).

kairstenfay replied:
I switched to using SmartyStreets' US Extract API for addresses where the normal lookup fails.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
+
+    url = 'https://us-street.api.smartystreets.com/street-address?'
+    url += f'auth-id={AUTH_ID}&auth-token={AUTH_TOKEN}&street='
+    url += f'{address}&candidates={NUM_CANDIDATES}'
```
The interpolated values aren't URL-encoded, which means they'll break if the address (or id/token) contains a reserved character, such as in `4th & Walnut`. Note that `requests` supports ways to pass a dict to generate query or body params, but also that this function won't be necessary if we use the SmartyStreets SDK.

kairstenfay replied:
Resolving in favor of using SmartyStreets SDK.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
+    df['response'] = df['address'].apply(lookup_address)
+    df['standard_address'] = df['response'].apply(response_to_standard_address)
+    df['latlng'] = df['response'].apply(response_to_latlng)
+    df['final_tract'] = df['latlng'].apply(latlng_to_polygon)
```
Latitude and longitude should end up as separate columns in the CSV.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,194 @@
 
-def load_geojson(geojson_filename):
-    """Read GeoJSON file and return a list of features converted to shapes."""
+    To minimize cost, a response for each address is only looked up once.
```
This claim in the docstring makes me think that the function is employing caching/memoization. The comment is more applicable to the _callers_ of this function, which are what control how many times an address is looked up.

kairstenfay commented:
Please see the new changes I've implemented per the previous code review.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,167 @@
+        raise ValueError(f"Unknown file extension for file named «{argv[1]}». \
+Please choose from one of the following file extensions: \n \
+    * .csv \n \
+    * .json")
```
FYI only, not necessary to change:

It can sometimes be nicer to write multi-line strings using Python's triple-quote syntax and [the `dedent` function from the textwrap module](https://docs.python.org/3/library/textwrap.html#textwrap.dedent). For example:

```python
        raise ValueError(dedent(f"""
            Unknown file extension for file named «{argv[1]}».
            Please choose from one of the following file extensions:
                * .csv
                * .json
        """))
```

kairstenfay replied:
Good to know! I'll use that from now on.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,167 @@
+        response = lookup_address(address)
 
-        if latlng:
+        if response and response['lat']:  # assume 'lng' will be present too
```
Why assume and not test?

kairstenfay replied:
Blind faith? I'll add that second test.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,167 @@
+        result = extract_address(credentials, address)
+
+    first_candidate = result[0]
+    return {"standard_address": first_candidate.delivery_line_1 + ' ' + 
```
What about `delivery_line_2`?

kairstenfay replied:
I wasn't sure if we needed that level of granularity, but I can easily add it back in.

tsibley commented on address_to_census_tract.py:
```diff
@@ -7,81 +7,167 @@
+        client.send_lookup(lookup)
+    except exceptions.SmartyException as err:
+        print(err)
+        return
```
Sending the error to stdout means it'll be mixed in with the CSV or JSON output.

kairstenfay replied:
Is this better handled by
`LOG.exception(err)` ?

tsibley commented on .gitignore:
```diff
@@ -7,3 +7,5 @@ environment*
 
 *.DS_Store
+
+/__pycache__
```
This change should really be in a separate commit from the README updates. The leading slash should also be removed so that `__pycache__` directories anywhere are ignored, not just the top-level one. Cached compiled Python modules should never be committed to version control.

tsibley commented on README.md:
```diff
@@ -2,57 +2,61 @@
-1. Geocode a given single-line address to a (latitude, longitude) pair using a
-   remote (forward) geocoding service.
+1. Geocode a given single-line address to a (latitude, longitude) pair using a remote (forward) 
+   geocoding service.
```
A lot of the changes to this README are re-wrapping lines from ~80 chars to a longer length. Doing so while making content changes makes the diff noisy and hard to read.

The actual content changes should also be part of the first commit on the branch.

tsibley commented on address_to_census_tract.py:
```diff
@@ -4,84 +4,194 @@
+    if argv[1].endswith('.json'):
+        process_json(argv[1])
+    elif argv[1].endswith('.csv'):
+        process_csv(argv[1])
```
The number of args should be checked before accessing `argv[1]`. Perhaps we should switch to using argparse now.

tsibley commented on address_to_census_tract.py:
```diff
@@ -4,84 +4,194 @@
+    Given a *file_path* to a CSV, processes the `address` column and adds 
+    extra columns for a standardized address, latitude and longitude
+    coordinates, and census tract. If a given address is invalid, these new 
+    columns are left blank. Saves the new table at `data/test/out.csv`. 
```
How about outputting the CSV to stdout instead of a hardcoded test file name? (Happy to explain why I think this preferred.)

kairstenfay replied:
I understand why a hardcoded test file name is not preferred, but why to stdout? Would you rather save the data via
`python address_to_census_tract.py {input_file} > {output_file}`
?

tsibley replied:
The convention of Unix command-line tools is to take input on stdin or named files and send output to stdout or named files.

I suggested stdout because the JSON output is already sent to stdout, and shell redirection is an easy way to save the output. Adding an `--output / -o` option would also be good to name the output file explicitly, but that doesn't have to happen in this PR unless you want to.

kairstenfay replied:
I like that idea. I'll switch to stdout for now and incorporate output options into the next PR.

kairstenfay committed:
Update gitignore

Ignore cached compiled Python modules

kairstenfay committed:
Add test CSV data

kairstenfay committed:
Use SmartyStreets geocoding API for addresses

These changes swapped Google's geocoding API for SmartyStreets' via its 
Python SDK. Using this service requires an authentication ID and token not tracked
through git that users need to add to their environment variables. 

The program currently works by passing a free text address via the `street`
parameter to the SmartyStreets  US Street API. If no match is found, the program 
re-tries the same address by passing it to the `text` parameter of the 
SmartyStreets US Extract API. The data is printed to stdout with additional
information on successful matches containing standardized address, latitude,
longitude, and census tract.

Currently, this program can be run with the following command.
`./address_to_census_tract.py {filepath}` where `filepath` has an extension of
either `.json` or `.csv`. See the example data sets in `data/test` for column or
category naming conventions.

kairstenfay commented:
I fixed the diffs so they're much more intelligible! Also, I'm now dumping the CSV to stdout and checking for the length of arguments before processing the input file.
